src/datamodel/validator/UUri.cpp

```diff
@@ -10,3 +10,179 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "up-cpp/datamodel/validator/UUri.h"
+
+namespace uprotocol::datamodel::validator::uri {
+
+using namespace uprotocol;
+
+std::string_view message(Reason reason) {
+	switch (reason) {
+		case Reason::EMPTY:
+			return "The URI authority_name is empty or whitespace.";
+		case Reason::RESERVED_VERSION:
+			return "The version is 0 (reserved).";
+		case Reason::RESERVED_RESOURCE:
+			return "The resource is 0 for non-RPC-response URIs.";
+		case Reason::DISALLOWED_WILDCARD:
+			return "The URI contains a wildcard in one or more fields that is "
+			       "not allowed for this form.";
+		case Reason::BAD_RESOURCE_ID:
+			return "The resource ID is not in the allowed range for this URI "
+			       "form.";
+		default:
+			return "Unknown reason.";
+	}
+}
+
+bool uses_wildcards(const v1::UUri& uuri) {
+	if (uuri.authority_name().find("*") != std::string::npos) {
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF) == 0xFFFF) {  // service ID
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF0000) == 0) {  // service instance ID
```

Comment:
Isn't this an `and` condition - the ID must be 0xFFFF _and_ the instance ID must be 0?

Reply:
I don't think so, the basics/uri.adoc section 5 "Pattern Matching" has 5 bullet points, and nothing says they depend on each other.

Also in section 5.1 there's a couple examples where the upper and lower halves behave independently.

Reply:
Ok, good enough for me :shrug: